File: frontend/src/services/portfolio.ts
import api from './api'
import type { Portfolio, Transaction, PortfolioValuation, WatchlistItem } from '../types'

export const portfolioService = {
  async list(): Promise<Portfolio[]> {
    const { data } = await api.get<Portfolio[]>('/portfolio/')
    return data
  },

  async create(name: string, description?: string): Promise<Portfolio> {
    const { data } = await api.post<Portfolio>('/portfolio/', { name, description })
    return data
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/portfolio/${id}`)
  },

  async addHolding(portfolioId: number, ticker: string, shares: number, avg_cost: number, sector?: string) {
    const { data } = await api.post(`/portfolio/${portfolioId}/holdings`, { ticker, shares, avg_cost, sector })
    return data
  },

  async removeHolding(portfolioId: number, holdingId: number): Promise<void> {
    await api.delete(`/portfolio/${portfolioId}/holdings/${holdingId}`)
  },

  async addTransaction(portfolioId: number, tx: {
    ticker: string; transaction_type: 'BUY' | 'SELL'; shares: number; price: number; notes?: string
  }): Promise<Transaction> {
    const { data } = await api.post<Transaction>(`/portfolio/${portfolioId}/transactions`, tx)
    return data
  },

  async getTransactions(portfolioId: number): Promise<Transaction[]> {
    const { data } = await api.get<Transaction[]>(`/portfolio/${portfolioId}/transactions`)
    return data
  },

  async getValuation(portfolioId: number): Promise<PortfolioValuation> {
    const { data } = await api.get<PortfolioValuation>(`/portfolio/${portfolioId}/valuation`)
    return data
  },

  async getWatchlist(): Promise<WatchlistItem[]> {
    const { data } = await api.get<WatchlistItem[]>('/portfolio/watchlist')
    return data
  },

  async addToWatchlist(ticker: string, notes?: string): Promise<WatchlistItem> {
    const { data } = await api.post<WatchlistItem>('/portfolio/watchlist', { ticker, notes })
    return data
  },

  async removeFromWatchlist(id: number): Promise<void> {
    await api.delete(`/portfolio/watchlist/${id}`)
  },
}

File: backend/app/models/portfolio.py
from sqlalchemy import Column, Integer, String, Float, ForeignKey, DateTime, Enum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from ..core.database import Base


class TransactionType(str, enum.Enum):
    BUY = "BUY"
    SELL = "SELL"


class Portfolio(Base):
    __tablename__ = "portfolios"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    description = Column(String)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)

    owner = relationship("User", back_populates="portfolios")
    holdings = relationship("Holding", back_populates="portfolio", cascade="all, delete-orphan")
    transactions = relationship("Transaction", back_populates="portfolio", cascade="all, delete-orphan")


class Holding(Base):
    __tablename__ = "holdings"

    id = Column(Integer, primary_key=True, index=True)
    portfolio_id = Column(Integer, ForeignKey("portfolios.id"), nullable=False)
    ticker = Column(String, nullable=False, index=True)
    shares = Column(Float, nullable=False)
    avg_cost = Column(Float, nullable=False)
    sector = Column(String)

    portfolio = relationship("Portfolio", back_populates="holdings")


class Transaction(Base):
    __tablename__ = "transactions"

    id = Column(Integer, primary_key=True, index=True)
    portfolio_id = Column(Integer, ForeignKey("portfolios.id"), nullable=False)
    ticker = Column(String, nullable=False)
    transaction_type = Column(Enum(TransactionType), nullable=False)
    shares = Column(Float, nullable=False)
    price = Column(Float, nullable=False)
    executed_at = Column(DateTime, default=datetime.utcnow)
    notes = Column(String)

    portfolio = relationship("Portfolio", back_populates="transactions")


class Watchlist(Base):
    __tablename__ = "watchlists"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    ticker = Column(String, nullable=False)
    added_at = Column(DateTime, default=datetime.utcnow)
    notes = Column(String)

    owner = relationship("User", back_populates="watchlists")
